File: BennuLib/IO/NativeFormatReader.vb
Option Infer On

Imports System.IO
Imports PkSprite.Core

Namespace Bennu.IO
    Public Class NativeFormatReader
        Inherits BinaryReader

        Private Shared ReadOnly _encoding As Text.Encoding = Text.Encoding.GetEncoding(850)

        Public Sub New(input As Stream)
            MyBase.New(input, _encoding)
        End Sub

        Public Function ReadDescription() As String
            Return _encoding.GetString(ReadBytes(32))
        End Function

        Public Function ReadMagic() As Magic
            ' 3 first bytes describe the depth of the MAP
            Dim fileType = _encoding.GetString(ReadBytes(3))
            ' Next 4 bytes are MS-DOS termination, and last is the MAP version
            Dim descriptor = ReadBytes(4)
            Dim version = ReadByte()

            Return New Magic(fileType, version, descriptor)
        End Function

        Public Function ReadPalette() As Byte()
            Dim paletteColors = ReadBytes(PaletteSize - 1)
            Return paletteColors
        End Function

        Public Function ReadPivotPoints(number As Integer) As PivotPoint()
            Dim points As New List(Of PivotPoint)
            For n As Integer = 0 To number - 1
                Dim point As New PivotPoint(n, ReadInt16, ReadInt16)
                ' Two coods set to -1 mean invalid point
                If Not (point.X = -1 And point.Y = -1) Then
                    points.Add(point)
                End If
            Next
            Return points.ToArray()
        End Function

        Public Function ReadUnusedPaletteGamma() As Byte()
            Return ReadBytes(ReservedBytesSize)
        End Function

        Public Function ReadPivotPointsNumber() As Integer
            Dim flags = ReadUInt16()
            Dim numberPivotPoints = Convert.ToInt16(flags And NumberOfControlPointsBitMask)
            Return numberPivotPoints
        End Function

        Public Function ReadGlyphInfo() As GlyphInfo
            Return New GlyphInfo(
                    ReadInt32,
                    ReadInt32,
                    ReadInt32,
                    ReadInt32)
        End Function
    End Class
End Namespace

File: BennuLib.Sharp/IO/NativeFormatReader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

using static BennuLib.IO.NativeFormat;

namespace BennuLib.IO
{
	public class NativeFormatReader : BinaryReader
	{

		private static readonly Encoding _encoding = Encoding.GetEncoding(850);

		public NativeFormatReader(Stream input) : base(input, _encoding)
		{
		}

        public string ReadAsciiZ(int length)
        {
            byte[] bytes = ReadBytes(length);

            int n = 0;
            for ( n = 0; n < length; n++ )
            { 
                if (bytes[n] == 0)
                    break;
            }

            string result;
            if (n > 0) {
                byte[] trimmedBytes = new byte[n];
                Array.Copy(bytes, 0, trimmedBytes, 0, n);
                result = _encoding.GetString(trimmedBytes);
            } else
            {
                result = "";
            }

            return result;
        }

		public Header ReadHeader()
		{
			// 3 first bytes describe the depth of the MAP
			var fileType = _encoding.GetString(ReadBytes(3));
			// Next 4 bytes are MS-DOS termination, and last is the MAP version
			var descriptor = ReadBytes(4);
			var version = ReadByte();

			return new Header(fileType, version, descriptor);
		}

		public byte[] ReadPalette()
		{
			var paletteColors = ReadBytes(NativeFormat.PaletteSize - 1);
			return paletteColors;
		}

		public PivotPoint[] ReadPivotPoints(int number)
		{
			List<PivotPoint> points = new List<PivotPoint>();
			for (int n = 0; n <= number - 1; n++) {
				PivotPoint point = new PivotPoint(n, ReadInt16(), ReadInt16());
				// Two coods set to -1 mean invalid point
				if (!(point.X == -1 & point.Y == -1)) {
					points.Add(point);
				}
			}
			return points.ToArray();
		}

		public byte[] ReadUnusedPaletteGamma()
		{
			return ReadBytes(NativeFormat.ReservedBytesSize);
		}

		public int ReadPivotPointsNumber()
		{
			var flags = ReadUInt16();
			var numberPivotPoints = Convert.ToInt16(flags & PivotPointsNumberBitMask);
			return numberPivotPoints;
		}

		public GlyphInfo ReadGlyphInfo()
		{
			return new GlyphInfo(ReadInt32(), ReadInt32(), ReadInt32(), ReadInt32());
		}
	}
}
